daemon/mgr/image.go

```diff
@@ -296,6 +303,80 @@ func (mgr *ImageManager) AddTag(ctx context.Context, sourceImage string, targetT
 	return err
 }
 
+// ImageHistory returns image history by reference.
+func (mgr *ImageManager) ImageHistory(ctx context.Context, idOrRef string) ([]types.HistoryResultItem, error) {
+	img, err := mgr.fetchContainerdImage(ctx, idOrRef)
+	if err != nil {
+		return nil, err
+	}
+
+	desc, err := img.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ociImage, err := containerdImageToOciImage(ctx, img)
+	if err != nil {
+		return nil, err
+	}
+
+	cs := img.ContentStore()
+	manifest, err := mgr.getManifest(ctx, cs, img, platforms.Default())
+	if err != nil {
+		return nil, err
+	}
+
+	ociImageHistory := ociImage.History
+	lenOciImageHistory := len(ociImageHistory)
+	if lenOciImageHistory < len(manifest.Layers) {
+		return nil, errors.New("number of image history info should be no less than number of manifest layers")
+	}
+
+	countNonEmptyLayer := 0
+	for _, h := range ociImageHistory {
+		if !h.EmptyLayer {
+			countNonEmptyLayer++
+		}
+	}
+	if countNonEmptyLayer != len(manifest.Layers) {
+		return nil, errors.New("number of non-empty layer in history info should equal to number of manifest layers")
+	}
+
+	history := make([]types.HistoryResultItem, lenOciImageHistory)
+	// Note: ociImage History layers info and manifest layers info are all in order from bottom-most to top-most, but the
+	// user-interactive history is in order from top-most to top-bottom, so we need to reverse ociImage History traverse order.
+	j := len(manifest.Layers) - 1
+	for i := range ociImageHistory {
+		history[i] = types.HistoryResultItem{
+			Created:    ociImageHistory[lenOciImageHistory-i-1].Created.UnixNano(),
+			CreatedBy:  ociImageHistory[lenOciImageHistory-i-1].CreatedBy,
+			Author:     ociImageHistory[lenOciImageHistory-i-1].Author,
+			Comment:    ociImageHistory[lenOciImageHistory-i-1].Comment,
+			EmptyLayer: ociImageHistory[lenOciImageHistory-i-1].EmptyLayer,
+			ID:         "<missing>",
+			Size:       0,
+		}
+
+		// TODO: here we just set imageID of top image layer, we do nothing with the lower image ID, after pouch
+		// enables build/commit functionality, we should get local lower image(parent image) layer ID.
+		if i == 0 {
+			history[i].ID = desc.Digest.String()
+		}
+
+		// Note: number of manifest layers should be less than ociImage History messages due to the existence of empty layers.
+		// The size of these empty layers should be set to 0.
+		if !history[i].EmptyLayer {
```

Comment:
we can do the valid check here, like
```
if j < 0 {
    return nil, fmt.Errorf("the layer data don't match the layers")
}
```

Out of the loop, we need to check the `j` is -1 or not. I think we can make the logic clean. How do you think?

Reply:
I think you are right.